### ISSUE TITLE
HTTPAuthSessionWrapper throws away first segment of path with anonymous requests

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3936]
|<img alt="esteve's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| esteve reported|
|-|-|
|Trac ID|trac#3936|
|Type|defect|
|Created|2009-07-29 09:44:43Z|

I introduced a bug in the way HTTPAuthSessionWrapper handles anonymous requests in [#3924](https://github.com/twisted/twisted/issues/3924) where the first segment of path is thrown away (see related bug [#3679](https://github.com/twisted/twisted/issues/3679)).

Attachments:

* [anonymous-httpauth-wrapper-child-3936.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ba1/ba1276a5b095d08a0c0f24ed7dd48b207132c62b/2167f2b5e1066a53a4cf725276604ec818f1c4cc.patch) (2089 bytes) - added by esteve on 2009-07-29 09:46:40Z - 
* [anonymous-httpauth-wrapper-child-3936-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ba1/ba1276a5b095d08a0c0f24ed7dd48b207132c62b/b5baff021820084418d23c75d22fcc352d2c74cb.patch) (4476 bytes) - added by esteve on 2009-09-01 12:17:29Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3936 3936
type__defect defect
reporter__esteve esteve
priority__normal normal
milestone__ 
branch__ 
branch_author__esteve esteve
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1248860683000000 1248860683000000
changetime__1251817020000000 1251817020000000
version__None None
owner__ 
cc__thijs cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="esteve's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|esteve removed owner|
|-|-|

This patch applies the same fix for [#3679](https://github.com/twisted/twisted/issues/3679) to anonymous requests (those that don't have an authorization header). I changed the HTTPAuthHeaderTests#test_getChildWithDefault and HTTPAuthHeaderTests#test_anonymousAccess test to reflect this change, and all tests pass. However, I had to use getChildForRequest for the first, instead of getChildWithDefault and I'm not sure if that's completely fine.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to esteve|
|-|-|

Using `getChildForRequest` in the tests should always be preferred.  That's how real requests are dispatched, so it ensures the tests are exercising the code in as realistic a way as possible.  There are a number of potential bugs which can be encountered when calling `getChildWithDefault` directly, due to various subtleties in the resource traversal APIs.  Arguably, the failure you encountered which caused you to make this change was one such.  So, hooray for that.

For the other test's changes, the previous version of that test was buggy, right?  It was asserting the wrong results in the response, because it was ignoring the fact that the `self.childName` child was being requested, and asserting that the response was rendered by the anonymous avatar directly, instead of its child?  If my understanding is correct, then this change looks good and I think the patch should be applied.  Please let me know if I've gotten it right and I'll apply the patch.

By the way, do you want commit access?

**Comment by twisted-trac** [comment]
|<img alt="esteve's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|esteve commented|
|-|-|

''Using getChildForRequest in the tests should always be preferred. That's how real requests are dispatched, so it ensures the tests are exercising the code in as realistic a way as possible. There are a number of potential bugs which can be encountered when calling getChildWithDefault directly, due to various subtleties in the resource traversal APIs. Arguably, the failure you encountered which caused you to make this change was one such. So, hooray for that.
''

Great! I wasn't sure because I don't know the details about getChildWithDefault's implementation and thought that the change was so trivial that the patch must be wrong :-)

_For the other test's changes, the previous version of that test was buggy, right? It was asserting the wrong results in the response, because it was ignoring the fact that the self.childName child was being requested, and asserting that the response was rendered by the anonymous avatar directly, instead of its child? If my understanding is correct, then this change looks good and I think the patch should be applied. Please let me know if I've gotten it right and I'll apply the patch._

The problem was that there was no way to attach a child resource to the UnprotectedResource, except by calling putChild. But then, I realized that the avatars dictionary already provided everything, sorry for adding that noise.

Anyway, we've been using this patch for a while and it works fine for us.

_By the way, do you want commit access?_

That's an honor that I'm sure I'm not worth of. Anyway, it also comes with a lot of responsibility, and I don't know if I'll always be able to contribute (sending patches, participating in the mailing list, reviewing patches, etc.) at the same rate. If you still think that's ok, it would be an honor to have commit access.

**Comment by twisted-trac** [comment]
|<img alt="esteve's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|esteve set status to `assigned`|
|-|-|

Can I help with this patch getting applied? I'll write more tests if you think it's necessary.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra set owner to @exarkun<br>@thijstriemstra set status to `new`|
|-|-|

Assigning it to exarkun.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to esteve|
|-|-|

I lost track of this ticket after my last comment.  I'd like to see `test_anonymousAccess` use something other than `self.avatar` as the anonymous avatar.  That way, the test really shows that anonymous login is getting the anonymous avatar; as it is now, the test will pass even if anonymous access were somehow implemented to get the "foo bar" avatar.

**Comment by twisted-trac** [comment]
|<img alt="esteve's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|esteve commented|
|-|-|

Replying to [exarkun](#note_6):
> I lost track of this ticket after my last comment.

No worries :-)

>  I'd like to see `test_anonymousAccess` use something other than `self.avatar` as the anonymous avatar.  That way, the test really shows that anonymous login is getting the anonymous avatar; as it is now, the test will pass even if anonymous access were somehow implemented to get the "foo bar" avatar.

I agree, I just did it to minimize the changes, but you're right that it's not too exhaustive. I changed the test to use a different resource in case an anonymous request comes in.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

This looks good now.  Thanks.  I'll apply it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

(In [27281]) Apply anonymous-httpauth-wrapper-child-3936-2.patch 

Author: esteve
Reviewer: exarkun
Fixes: [#3936](https://github.com/twisted/twisted/issues/3936)

Change two of the `HTTPAuthSessionWrapper` tests to assert that the rendering
result came from the requested child, rather than from the avatar itself.
These tests were previously asserting the wrong behavior.  Also fix a bug
this revealed in the implementation of the anonymous login support.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
